### PR TITLE
[KED-2528] Improve layer solving approach when layers partially defined

### PR DESCRIPTION
**Pull request by bru5**
## Description

This PR improves the approach for inferring and rendering nodes in pipelines with only partially defined layers.

## Development notes

The algorithm for nearest layer search was changed to strictly follow rank order, avoiding edge cases where layer constraints could not be solved with the previous algorithm.

The strict constraint solver was also changed to allow partial solving whilst throwing a console warning, though this was no longer observed to happen in practice.

## QA notes

Tested using a large pipeline that has only partially defined layers (i.e. not the 'demo').
The 'demo' dataset, which has fully defined layers, was also tested.

Using both cases above, open the console to observe any errors or warnings, then test by enabling layers and toggling various combinations of nodes as well as toggling top level pipelines in various combinations. 

Each time check that the resulting render is in a sensible order, does not appear broken and there are no console errors or warnings. Since the order is inferred using only limited information, it may not always be exactly as desired but in this case the idea is that the user should add more layer information to nodes.

## Checklist

- [x] Read the [contributing](/CONTRIBUTING.md) guidelines
- [ ] Opened this PR as a 'Draft Pull Request' if it is work-in-progress
- [ ] Updated the documentation to reflect the code changes
- [ ] Added new entries to the `RELEASE.md` file
- [ ] Added tests to cover my changes

## Legal notice

- [x] I acknowledge and agree that, by checking this box and clicking "Submit Pull Request":

- I submit this contribution under the [Apache 2.0 license](https://www.apache.org/licenses/LICENSE-2.0.txt) and represent that I am entitled to do so on behalf of myself, my employer, or relevant third parties, as applicable.
- I certify that (a) this contribution is my original creation and / or (b) to the extent it is not my original creation, I am authorised to submit this contribution on behalf of the original creator(s) or their licensees.
- I certify that the use of this contribution as authorised by the Apache 2.0 license does not violate the intellectual property rights of anyone else.

### CODE REVIEW

File: src/utils/graph/graph.js [hunk 1]
```diff
@@ -77,39 +77,41 @@ export const addEdgeLinks = (nodes, edges) => {
...
+    for (const layer of layers) {
+      validLayers[layer] = true;
+    }
+
+    const hasValidLayer = (node) => Boolean(node?.layer in validLayers);
+    const lastLayer = layers[layers.length - 1];
 
+    // For each node
     for (const node of nodes) {
+      // Find first descendant node that has a valid layer following rank order (including itself)
```
Comment (datajoely) [inline src/utils/graph/graph.js:97]: What are they ranked by?

File: src/utils/graph/solver.js [hunk 2]
```diff
@@ -75,15 +75,25 @@ const solveStrict = (constraints, constants) => {
...
+          variables[variableId(co.b, co.base.property)]
+        )
+      );
+    } catch (err) {
+      unsolvableCount += 1;
+    }
+  }
+
+  if (unsolvableCount > 0) {
+    console.warn(`Skipped ${unsolvableCount} unsolvable constraints`);
```
Comment (datajoely) [inline src/utils/graph/solver.js:96]: Is there no more information worth spitting out to the console?

File: src/utils/graph/graph.js (hunk 1 above)
Comment (limdauto) [reply src/utils/graph/graph.js:97]: Nodes are ranked by their batch-topological order

File: src/utils/graph/solver.js (hunk 2 above)
Comment (limdauto) [reply src/utils/graph/solver.js:96]: Yea, would be nice to improve this error message some how

**Review by limdauto — APPROVED** [review]
Thank you 🙏  this does make sense.

File: src/utils/graph/solver.js (hunk 2 above)
Comment (bru5) [reply src/utils/graph/solver.js:96]: So I did experiment with this, but essentially the problem is that the constraint systems involved are not very human interpretable given their size and complexity, such that the individual constraints here themselves don't really directly relate to the cause of the user's problem in a meaningful way.

I think for now the notification will be enough to keep an eye on given we're not expecting to see it after this fix, but if there are further complications still then we can invest more into somehow making this more interpretable.

Comment (datajoely) [reply src/utils/graph/solver.js:96]: Could we at least put the notification on screen rather than in the console?

**Pull request merged**